src/python/pants/backend/javascript/subsystems/nodejs.py

```diff
@@ -441,10 +523,35 @@ async def determine_nodejs_binaries(
     return NodeJSBinaries(os.path.dirname(paths_per_version[satisfying_version][0].path))
 
 
+async def prepare_default_package_managers(
+    environment: NodeJSProcessEnvironment, nodejs: NodeJS
+) -> list[str]:
+    version_tuples = sorted(
+        f"{pkg_manager}@{version}" for pkg_manager, version in nodejs.package_managers.items()
+    )
+    if version_tuples:
+        await Get(
+            ProcessResult,
+            Process(
+                argv=("corepack", "prepare", *version_tuples, "--activate"),
+                description=f"Preparing default configured package managers {', '.join(version_tuples)}.",
+                immutable_input_digests=environment.immutable_digest(),
+                level=LogLevel.DEBUG,
+                env={**environment.to_env_dict()},
+                append_only_caches={**environment.append_only_caches},
+            ),
+        )
+    return version_tuples
+
+
 @rule(level=LogLevel.DEBUG)
 async def setup_node_tool_process(
-    request: NodeJSToolProcess, environment: NodeJSProcessEnvironment
+    request: NodeJSToolProcess, environment: NodeJSProcessEnvironment, nodejs: NodeJS
 ) -> Process:
+    versions = await prepare_default_package_managers(environment, nodejs)
+    pkg_manager_env = {
+        "__PACKAGE_MANAGER_VERSIONS": ",".join(versions)
+    }  # Invalidates cached process in event of update
```

Comment:
Is this abuse of append_only_caches? I noticed processes were not re-run after I changed the npm version, which is very dangerous if a user is attempting an upgrade.

Reply:
I think I've convinced myself that yes, this is very bad (and I think its the reason CI is failing).

Reply:
https://github.com/pantsbuild/pants/pull/18748/commits/500b293c97536bb2e7a7a61bccfd2daf9c4739b8